### DISCUSSION TITLE
How to force dag to run like a simple cron job?

### DISCUSSION THREAD

**Discussion by RafayAK** [apache/airflow#15432] [Q&A] [answered]
Airflow scheduler kinda left me scratching my head for the past few days as it backfills dag runs even after `catchup=False`.
My timezone-aware dag has a start date of `13-04-2021 19:30 PST` or `14-04-2021 2:30 UTC`  has the following configuration:

```python
# define DAG and its parameters
dag = DAG(
    'backup_dag',
    default_args=default_args,
    start_date=pendulum.datetime(2021, 4, 13, 19, 30, tz='US/Pacific'),  # set start_date in US/Pacific (PST) timezone
    description='A data backup pipeline',
    schedule_interval="30 19 * * *",  # 7:30 PM every day
    catchup=False,
    is_paused_upon_creation=False
)
```

This dag runs on an edge device, that edge device is sometimes on and sometimes off. I want this dag to basically schedule its run at `19:30 PST` or `2:30 UTC`, whenever the edge device is on, otherwise skip. The weird thing is that when I deploy the container with the dag to the edge device the dag automatically starts its first run outside the scheduled interval, even though that interval has passed!

![Screenshot from 2021-04-19 16-50-54](https://user-images.githubusercontent.com/17025191/115232905-ad202800-a130-11eb-8e2e-7c08ad1bef90.png)

What am I missing here? I can't wrap my head around why the scheduler is doing this

Following is my understanding after reading all the documentation, please do correct me if I'm wrong.

DAG picked up by scheduler at `2021-04-19T011:30:00+00:00 UTC`, ideally it should run at `2021-04-20T02:30:00+00:00 UTC` according to the dag config. All times below are in UTC

```
      Dag Start_date		 1st run(skip catchup=false)   2nd run(skip catchup=false)    3rd run(skip catchup=false)   4th run(skip catchup=false)
2021-04-14T02:30:00+00:00 ---> 2021-04-15T02:30:00+00:00 ---> 2021-04-16T02:30:00+00:00  ---> 2021-04-17T02:30:00+00:00 ---> 2021-04-18T02:30:00+00:00 ---> 

5th run(skip catchup=false)   7th run(should execute)              
 2021-04-19T02:30:00+00:00 ---> 2021-04-20T02:30:00+00:00

```

So, why is the 5th run taking place for interval `2021-04-18T02:30:00+00:00` to `2021-04-19T02:30:00+00:00` even though the interval has passed?

I want the DAG to only run when its interval has come.

**Comment by RafayAK** [comment]
I have been reading about `LatestOnlyOperator`, will setting that as the first task in the dag fix this problem and make it behave more like a simple cron job?

**Reply by RafayAK** [reply]
`LatestOnlyOperator` does not work. 😐

**Comment by RafayAK** [comment]
Seems like I've found a solution and that is to set the `start_date` dynamically. 

`start_date= pendulum.yesterday('US/Pacific').add(hours=19, minutes=30)`

Don't know if this is the right solution though. Airflow best practices often mention against using a dynamic start date.

**Reply by RafayAK** [reply]
This doesn't always work, don't know why though.

**Comment by RafayAK** [comment, marked as answer]
Asked this question on StackOverflow:
https://stackoverflow.com/questions/67212884/how-to-make-a-dag-in-apache-airflow-run-like-a-simple-cron-job/67218114?noredirect=1#comment118816954_67218114

Got an answer that, suggests to stop scheduling the task with airflow and use cron to externally executed the dag.

**Comment by RafayAK** [comment]
Tried the method as suggested in my StackOverflow Question. Works fine for now. Kinda a surprise that such a use case is not supported by default and requires adding additional dependencies out side of AIrflow for orchestration.

**Reply by potiuk** [reply]
There is a change coming (likely in 2.2) that addresses this: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval . The main reason was that Airflow is originaly designed to process "data intervals" (so batches of data) where processing happened "after" the interval finished. That's why it currently runs at the END of interval by default. The "Richer schedule intervals" feature aims at adding more flexibility for both - defininig the schedules (you will be able to define it in plain cron-like fashion or have python functions that allow for custom calendars, sunset/sunrise and other kinds of schedules) as well as it allows to choose if you want to run in "data interval" fashion or "cron fashion".

Stay tuned for the announcements at the devlist and user list.

**Reply by RafayAK** [reply]
can't wait

**Reply by RafayAK** [reply]
I'll update the findings after the new update here as well as my StackOverlfow question

**Reply by mai-nakagawa** [reply]
[The mentioned document](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-39+Richer+scheduler_interval) introduces `DataTimetable` and `CronTimetable` classes - the diff is whether a DAG Run is kicked off at the start of the period or immediately after the DAG is added. However, Airflow 2.2 actually released only with `CronDataIntervalTimetable` class which achieves the default behavior - DAG Run starts right after.

So, we need to [customize DAG scheduling with Timetables](https://airflow.apache.org/docs/apache-airflow/stable/howto/timetable.html#) to achieve the plain cron-like DAG scheduling. Correct me if I'm wrong.

**Reply by mai-nakagawa** [reply]
I just created a PR to add a timetable for the normal cron behavior: https://github.com/apache/airflow/pull/23662
